### ISSUE TITLE
Issues plotting an image

### ISSUE THREAD

**Issue by logankilpatrick** [jkrumbiegel/MakieLayout#25]
Hey @jkrumbiegel ! 

I am having a weird issue. When I plot an image, the bounds are correct: 1600 x 1600

However, when I zoom in or plot a sub-image, the coordinates from the image on my computer, let's say (1500, 700), do not show the same location. 

For some reason, the location in the makie window shows up 150 pixels up on the y-axis from where I would expect it to be. 

I am using `las[3, 1].aspect = DataAspect()`. 

Any thoughts as to what could be causing this issue?

**Comment by jkrumbiegel** [comment]
Can you post a quick MWE and a screenshot of the problem? I don’t quite understand yet :)

**Comment by logankilpatrick** [comment]
Here is what it looks like with plots :
<img width="316" alt="Screen Shot 2020-01-30 at 8 25 05 AM" src="https://user-images.githubusercontent.com/35577566/73453494-20762800-433a-11ea-8a71-5d2439c00439.png">

Here is what it looks like with MakieLayout at the same location: 
<img width="478" alt="Screen Shot 2020-01-30 at 8 25 23 AM" src="https://user-images.githubusercontent.com/35577566/73453520-3257cb00-433a-11ea-8414-98651e67470f.png">

**Comment by logankilpatrick** [comment]
The only non-standard thing I am doing is `img = rotr90(img, 1) #rotate right 90 degrees`. Otherwise, I am using basic image code: 

```jl
scene = Scene(resolution = (1000, 1000));
    screen = display(scene)
    campixel!(scene);

    maingl = GridLayout(scene, alignmode = Outside(30))

    ni = 2
    nj = 2

    las = maingl[1:ni, 1:nj] = [LAxis(scene, titlesize=20) for i in 1:ni, j in 1:nj]

    img = load("path to image")
    img = rotr90(img, 1) #rotate right 90 degree's

    image!(las[1, 1], img)
    image!(las[1, 2], img)
    image!(las[2, 1], img)
    image!(las[2, 2], img)
```

**Comment by jkrumbiegel** [comment]
That's not an example of the error, though :) First, I don't understand what the problem in the upper image is because there's no code for that (the y axis is different but I don't know how you create the plot) and I don't know what the problem with the lower code is (if there is any) because there is no visual for that, and I don't have your image file.

**Comment by logankilpatrick** [comment]
![zone-map](https://user-images.githubusercontent.com/35577566/73456175-6e415f00-433f-11ea-871f-a3b90b55761d.png)

**Comment by logankilpatrick** [comment]
If you just open up the image (in preview for me on macOS), and go to (1520, 710) like the axis's in the plots.jl image, you will see the same thing. But when MakieLayout one is different.

**Comment by logankilpatrick** [comment]
To summarize, the plots.jl image is what I am going after. When I plot the same image in MakieLayout, the `y` axis is weirdly altered.

**Comment by logankilpatrick** [comment]
The plots code is: `Plots.plot([f.x_min, f.x_max], [f.y_min, f.y_max], zone_img[f.y_min:1:f.y_max, f.x_min:1:f.x_max], yflip = true, legend = false, grid = false)` which is the end goal of what I was trying to do with MakieLayout, but the location it zooms in on in Makie when I do this: ` image!(zoneScene, [ymin, ymax], [xmin, xmax], zoneimage[ymin:1:ymax, xmin:1:xmax])` is different than the plots one so I reverted back to just plottting the image without constrining the x or y axis.

**Comment by jkrumbiegel** [comment]
and why do you pass the yrange first and then the xrange? shouldn't it be the other way around? maybe you're just cutting out a different part of the image

**Comment by jkrumbiegel** [comment]
The y location of the yellow spot in your Plots image is ca. 705 while in Makie it's 895, since 705 + 895 = 1600 this suggests that your y axis is flipped somehow in your processing pipeline, even though the orientations of both images come out the same

**Comment by logankilpatrick** [comment]
My mistake, I was on the master branch of my code and not the layout branch :), I am doing this: 

   `image!(las[1, 3], [x_min, x_max], [y_min, y_max], img[x_min:1:x_max, y_min:1:y_max])`

**Comment by logankilpatrick** [comment]
But the result is the same, still shows the y-axis offset by ~150 from where I would expect it to be.

**Comment by jkrumbiegel** [comment]
how are x_min etc determined?

**Comment by logankilpatrick** [comment]
The short story is that there's a bunch of simulations going on in the background and then there's a traverse plotted onto those images and the min/max are calculated via the location of the traverse +/- some padding.

**Comment by jkrumbiegel** [comment]
so in Plots you do
```julia
zone_img[f.y_min:1:f.y_max, f.x_min:1:f.x_max]
```
while in Makie you rotate the image to the right first and then do

```julia
img_rot90[x_min:1:x_max, y_min:1:y_max]
```

These are different, no?

**Comment by logankilpatrick** [comment]
Plots displays my images correctly to start whereas Makie and subsequently, MakieLayout does not hence the need to rotate the image. This works fine in Makie but not in MakieLayout. The difference is that in Makie I do `Makie.rotate!(zoneScene, -0.5pi)` rather than `img = rotr90(img, 1)`. This is likely the root of the issue.

**Comment by jkrumbiegel** [comment]
yes you need to rotate the indices correctly as well, not just flip the axes. I think you inadvertently flipped your y indices in the process, hence the wrong cutout

**Comment by logankilpatrick** [comment]
How does one flip the indices?

**Comment by jkrumbiegel** [comment]
can you tell me the full range of values that you want to cut out? you kept referring to 1520, 710 but that's only a point

**Comment by jkrumbiegel** [comment]
this shows again the value of a good MWE, as I'm missing the critical variables here and have to guess what's going on ;)

**Comment by jkrumbiegel** [comment]
why the images in makie need 90 degrees rotation I don't know, maybe because Julia uses column-major and OpenGL doesn't or something like that

**Comment by logankilpatrick** [comment]
Ranges: 
```
xmin: 1482
xmax: 1572
ymin: 666
ymax: 756
```

**Comment by logankilpatrick** [comment]
But these values change constantly on each step/plot of the program.

**Comment by jkrumbiegel** [comment]
by rotating the indices I mean that rotating the image 90 degrees to the right causes x to turn into flipped_y and y into x. that's why you don't index with y, x but with flipped_y, x

**Comment by logankilpatrick** [comment]
ah, well that's a headache. Is there a way to transpose the image such that I can stick with default x and y ?

**Comment by jkrumbiegel** [comment]
this is always a bit confusing with different plotting software, for example Plots also needs yflip = true in order to produce your desired result, so it's not like it immediately shows the correct result. But once you get it right it should be simple for every other plot

**Comment by logankilpatrick** [comment]
Yeah I use `yflip = true` in my plots code

**Comment by jkrumbiegel** [comment]
yeah so your ranges refer to the original image right?

**Comment by logankilpatrick** [comment]
yes

**Comment by logankilpatrick** [comment]
The reason I sort of need to stick with those number is that I do a line plot which has values between those original numbers on top of the image

**Comment by jkrumbiegel** [comment]
Haha ok I solved it, so your problem is that you index into the rotated image, but that messes with your indices. If you accept that plotting a rotr90 image is the same result in Makie as plotting a yflip = true image in Plots then you just need to keep everything the same before actually plotting. So this works: 

```julia
function test1()
    scene, maingl = layoutscene(resolution = (1000, 1000))

    ni = 1
    nj = 2

    las = maingl[1:ni, 1:nj] = [LAxis(scene, titlesize=20) for i in 1:ni, j in 1:nj]

    tightlimits!.(las)

    img = load("the image path")

    img_cutout = img[ymin:ymax, xmin:xmax] # index like with Plots

    image!(las[1], rotr90(img))
    image!(las[2], xmin:xmax, ymin:ymax, rotr90(img_cutout)) # only rotate before showing


    las[1].aspect = DataAspect()
    las[2].aspect = DataAspect()

    scene
end
test1()
```
<img width="801" alt="grafik" src="https://user-images.githubusercontent.com/22495855/73461309-3acff600-437a-11ea-82db-0df7ad980815.png">

**Comment by logankilpatrick** [comment]
Awesome, thanks! That does the job. I have to plot on that image now, and I want to do the following which works in Makie: 

```jl
positions = lift(stepslider[end][:value]) do val2
        Point2f0.(x_vec[1:val2], y_vec[1:val2])
end

    #This is what we will use to fix the "Patchy plotting"...
    function xlines(scene, points; kw...)
        l = lift(points) do points
            result = Point2f0[]
            for i in 1:(length(points)-1)
                push!(result, points[i], points[i + 1])
            end
            result
        end
        Makie.linesegments!(scene, l; kw...)
    end

xlines(las[1, 2], positions, overdraw=true, linewidth = 2)

```
This doesn't give any errors or anything like that but the orientation is incorrect.

**Comment by logankilpatrick** [comment]
<img width="764" alt="Screen Shot 2020-01-30 at 11 36 54 AM" src="https://user-images.githubusercontent.com/35577566/73469723-d5b5d980-4354-11ea-97ec-4970e560c7b6.png">
 this is what the orientation looks like.

**Comment by jkrumbiegel** [comment]
I think you might still be a bit confused with your coordinate transformations :) I believe Makie is plotting what you give it correctly, but in Plots you have yflip = true, so that flips your data like you see it here. In Makie there's no flip, so everything's upside down. Can you check if the y-values in your vector describe a ⋃ or a ⋂? If they do a ⋂ then Makie is actually "correct" while Plots is flipped. You have to really make sure that you understand what your x and y dimensions mean and how they should be oriented relative to the image / line plots.

**Comment by jkrumbiegel** [comment]
At least the indexing into the image like `img[ymin:ymax, xmin:xmax]` looks fishy to me, although I see no natural direction in the image so I can't say whether it is oriented correctly in my last solution, it is simply oriented like the Plots plot, but maybe that one is already incorrect because of your indexing.

**Comment by logankilpatrick** [comment]
I see what you are saying. In this case, the source of truth is the plots image. I just need to figure out how to flip things on Makie's end. 

This works in pure Makie because I rotate the whole scene, and then plot the lines on it. But in this case, I am not rotating the whole scene, I am just rotating the image with MakieLayout. 

Any suggestions on a way to get `positions` rotated by "-0.5pi"?

**Comment by jkrumbiegel** [comment]
> because I rotate the whole scene, and then plot the lines on it

That seems like a weird way of doing it, somebody reading your code would probably have a lot of trouble to understand what is supposed to end up where.

> Any suggestions on a way to get positions rotated by "-0.5pi"?

```julia
Base.rotr90(p::T) where T <: Point2 = T(p[2], -p[1])
Base.rotl90(p::T) where T <: Point2 = T(-p[2], p[1])
Base.rot180(p::T) where T <: Point2 = T(-p[1], -p[2])
```

**Comment by logankilpatrick** [comment]
Do you know how to rotate an array with floats? It seems to want ints.

**Comment by jkrumbiegel** [comment]
what do you mean, a 2D array with floats? this works for me
```julia
rotl90([[1.0 2.0]; [3.0 4.0]])
```

**Comment by logankilpatrick** [comment]
My bad, I need to rotate positions which is as follows:

```jl
positions = lift(stepslider[end][:value]) do val2
        Point2f0.(x_vec[1:val2], y_vec[1:val2])
    end
```

The positions variable is `::Observables.Observable{Array{GeometryTypes.Point{2,Float32},1}}`

**Comment by logankilpatrick** [comment]
So I'm not sure how to rotate it since it's not a 2D array

**Comment by asinghvi17** [comment]
`rotr90.(Point2f0.(...))`
Edit: You want to avoid doing that because it instantiates 2 Point2f0 arrays.  You could just do it manually by changing the call inside lift to `Point2f0.(y_vec[1:val2], -x_vec[1:val2])`

**Comment by logankilpatrick** [comment]
<img width="452" alt="Screen Shot 2020-01-30 at 3 53 09 PM" src="https://user-images.githubusercontent.com/35577566/73489474-a31dd800-4378-11ea-9116-bd5690c6842b.png">

**Comment by logankilpatrick** [comment]
This makes the plots tiny and moves them to the top right corner

**Comment by jkrumbiegel** [comment]
> `rotr90.(Point2f0.(...))`
> Edit: You want to avoid doing that because it instantiates 2 Point2f0 arrays. You could just do it manually by changing the call inside lift to `Point2f0.(y_vec[1:val2], -x_vec[1:val2])`

does it really? it should fuse the calls into one

**Comment by jkrumbiegel** [comment]
> This makes the plots tiny and moves them to the top right corner

Can you give another example with some minimal code that creates that problem for you? I can't help you well with an image alone.

**Comment by logankilpatrick** [comment]
I think this should do it: 

```jl
x_vec = [1525.77173, 1525.77173, 1525.77173, 1525.77173]
y_vec =  [709.3146, 709.3146, 709.3146, 709.3146]
x_min = 1511
x_max = 1541
y_min = 694
y_max = 724

scene, maingl = layoutscene(resolution = (1500, 1500))
maingl = GridLayout(scene, alignmode = Outside(30))

nrows = 2
ncols = 3

las = maingl[1:nrows, 1:ncols] = [LAxis(scene, titlesize=20) for i in 1:nrows, j in 1:ncols]
tightlimits!.(las)

#TODO make this work with the new MakieLayout Slider function. 
stepslider = slider(1:length(x_vec), "Steps", raw = true, camera = campixel!, start = length(x_vec))

positions = lift(stepslider[end][:value]) do val2
    Point2f0.(x_vec[1:val2], y_vec[1:val2])
end

#This is what we will use to fix the "Patchy plotting"...
function xlines(scene, points; kw...)
    l = lift(points) do points
        result = Point2f0[]
        for i in 1:(length(points)-1)
            push!(result, points[i], points[i + 1])
        end
        result
    end
    Makie.linesegments!(scene, l; kw...)
end

img = load("path to file")

image!(las[1, 1], rotr90(img))
image!(las[1, 2], x_min:x_max, y_min:y_max, rotr90(img[y_min:y_max, x_min:x_max])) # only rotate before showing
image!(las[1, 3], x_min:x_max, y_min:y_max, rotr90(img[y_min:y_max, x_min:x_max]))

image!(las[2, 2], x_min:x_max, y_min:y_max, rotr90(img[y_min:y_max, x_min:x_max]))
image!(las[2, 3], x_min:x_max, y_min:y_max, rotr90(img[y_min:y_max, x_min:x_max]))

las[1, 1].title = "a"
las[1, 2].title = "b"
las[1, 3].title = "c"

las[2, 1].title = "x"
las[2, 2].title = "y"
las[2, 3].title = "z"

las[1, 1].aspect = DataAspect()
las[1, 2].aspect = DataAspect()
las[1, 3].aspect = DataAspect()

las[2, 2].aspect = DataAspect()
las[2, 3].aspect = DataAspect()


xlines(las[1, 2], positions, overdraw=true, linewidth = 2)
xlines(las[1, 3], positions, overdraw=true, linewidth = 2)
xlines(las[2, 2], positions, overdraw=true, linewidth = 2)
xlines(las[2, 3], positions, overdraw=true, linewidth = 2)

display(scene)
```

**Comment by jkrumbiegel** [comment]
if I run that code, `positions` contains:
```julia
Observable{Array{Point{2,Float32},1}} with 4 listeners. Value:
Point{2,Float32}[[1525.7717, 709.3146], [1525.7717, 709.3146], [1525.7717, 709.3146], [1525.7717, 709.3146]]
```

So it's expected that this would be a tiny plot, because you're only plotting one point basically. I'm going to close this issue now because the original problem was solved and it becomes hard to follow, if there's a specific new problem that you have just open a separate one for that :)
